### ISSUE TITLE
Better manage Zim modules and prompts allowing the project to grow in an organized way

### ISSUE THREAD

**Issue by Eriner** [zimframework/zim#2]
_From @ericbn on September 28, 2016 16:43_

I believe Oh My Zsh and even Prezto failed to grow in an organized fashion mainly because they lack incorporating some kind of plugin system that would allow people to contribute with external modules and keep the core clean and simple.

I suggest we take this step with Zim (and make it awesome)!

My suggestion would go in the line of:
- Creating a `zim` GitHub organization, having the Zim core be `zim/zim`, and other Zim modules be like `zim/xyz`.
- Keeping into the Zim core only what we would agree are the basic modules.
- Allowing users to configure external prompts as easily as having this in `.zimrc`:

        zprompt_theme='subnixr/minimal'

- Allowing users to declare external modules also as easily as (two last ones):

        zmodules=(directory environment git history input utility meta custom \
                  syntax-highlighting history-substring-search prompt completion \
                  zim/pacman foo/bar)

Maybe we can incorporate an external existing plugin system into the Zim core to accomplish this, or we can implement our own system (which I think would be the more flexible way to go).

The plugin system would have at least functions like:
- `install` to install new declared modules
- `update` to update all existing modules
- `clean`/`cleanup` to remove unused modules that are still installed

In case we use an internal plugin system, that could be managed by `zmanage`, which we already have. (Side suggestion: wouldn't it be cooler to rename `zmanage` to `zim`?)

Maybe the `zim` organization could be the default for external modules, so if the user declares simply `pacman` inside `zmodules` and there's no such module, the plugin system would try to check for a `zim/pacman` repo to install from there.

I would not bother making the plugin system capable of importing Oh My Zsh or Prezto modules.

My suggestion of the basic modules for the Zim code are:
- All those importing the Zsh ones: `completion`, `history-substring-search`, `syntax-highlighting`
- All the ones configuring the sane Zsh defaults for all/most OSs: `directory`, `environment`, `history`, `input`, `spectrum`(?), `ssh` (only not useful for Mac OS X), `utility`
- The zsh core ones: `debug` and `meta`
- `git`, as git is needed to install and keep Zim and external modules updated
- `prompt` with a small selection of internal prompts (we have `eriner`, `gitster`, `magicmace`, `steeef`; there's already a newer fork of `minimal` as reported in https://github.com/Eriner/zim/issues/78 and we could check if that is the case with any of the others)

I'd suggest `archive`(?), `fasd` and `pacman` go as external modules inside the `zim` organization.

This would instantly make users use whatever external prompt they want, without thinking about creating a "Add xyz external prompt" pull request. Other module requests like https://github.com/Eriner/zim/issues/33 (`virtualenv`) or https://github.com/Eriner/zim/issues/82 (a module for prompts showing git info) would become candidates for external modules, and we even could decide whether we choose to adopt the module inside the `zim` organization or suggest the contributor to keep it as a personal project inside its own GitHub user account.

Thoughts?

_Copied from original issue: Eriner/zim#88_

**Comment by Eriner** [comment]
I've only read the first line of your proposal; this is something I realized quickly after starting. It would be ideal if all the modules had their own repository, and zim/zim would be strictly the "loader" and associated tasks. I'll read the rest of your proposal and respond at some point this weekend. This would be a fairly large undertaking and would require a fair amount of code refactoring, so I see this as a good long-term goal.

**Comment by Eriner** [comment]
_From @Plloi on September 28, 2016 19:25_

To maintain theme I propose callnowia or cbp (conveyor belt planet) as names for a utility for managing modules.

EDIT: Or possibly Vort.

**Comment by Eriner** [comment]
To follow this up, I have created an organization and will start to populate that sometime next week. In the meantime, any suggestions or discussion around this topic belongs here.

@ericbn:
 > Side suggestion: wouldn't it be cooler to rename `zmanage` to `zim`?

No, I don't want to clobber the zim-wiki binary name.

In terms of the module organization, all of the current modules would get their **own** repositories within the organization; zim itself will be _only_ the loader (basically `init.zsh`, save some helper scripts). This will make module management much easier, and will allow me to designate users as maintainers of specific modules. For example, I don't own a Mac, so if we wanted a Mac-specific module, I wouldn't be maintaining that. 

@Plloi:
 > To maintain theme I propose callnowia or cbp (conveyor belt planet) as names for a utility for managing modules.
 oh come now, if we need a meta tool to fit the theme, it's `GIR` (or possibly `SIR`).

Also, I will likely be making the move from submodules to subtrees in terms of zim-core management.

**Comment by Eriner** [comment]
_From @ericbn on September 28, 2016 22:53_

@Eriner, cool! I'd be happy to help coding whatever is needed.

**Comment by Eriner** [comment]
Thanks @ericbn, I (and I'm sure many other users) appreciate that, as well as the work you've done already! :) :+1:

**Comment by Eriner** [comment]
_From @Plloi on September 29, 2016 1:22_

Not sure i would trust GIR with something so important. Plus there is fun to be have with callnowia and cbp...

callnowia (order|return) for install/uninstall

or a megadoomer easter egg for cbp ...Megadoomer!

**Comment by Eriner** [comment]
_From @ericbn on September 29, 2016 2:20_

Just googled those names you guys are mentioning and found this [series](http://zim.wikia.com/wiki/Invader_Zim_(series)) I've never heard of (maybe because I don't watch Nickelodeon)... Is "Zim" really inspired in this series, and you seriously plan naming things after this?

**Comment by Eriner** [comment]
_From @Plloi on September 29, 2016 2:27_

Well the graphic in the readme is Zim's(or at least an Irken) head, inspiration acheived!

As for naming everything else well that's up to @Eriner, but it would make me happy because [I love this show](https://www.youtube.com/watch?v=wdMHgc2HzQU)

**Comment by Eriner** [comment]
I am going to migrate this issue to the new issue tracker, so follow the reference below if you're looking for the current discussion.

**Comment by Eriner** [comment]
I've uploaded the individual modules to the organization. Unfortunately, in terms of "upgrade" path (coming from Eriner/zim to zimframework/zim), I think the best option will be to **transfer the Eriner/zim repository to this organization**, then begin the refactor/restructure from there. This will allow users to simply `zmanage update` (which possibly should be renamed to `zmanage upgrade`) and immediately be connected to the new repository thanks to GitHub forwarding the old Eriner/zim URLs.

I think that is the best way to keep current users connected while making this transition. Unfortunately, I think that will mean killing off this repository and the associated issues. I should have thought about this more before creating this repository. Regardless, I will do my best to archive/transfer the issues from this repository back once the transfer is complete.

Before I make the transfer of the core repository, are there any objections to going this route?

CC: @ericbn @Tmplt @dvcrn @ezintz @stefanosc @rtfpessoa and any other's that have expressed interest in assisting in development and maintenance from the [is prezto dead?](https://github.com/sorin-ionescu/prezto/issues/1239) thread.

After we decide and make the initial transition, we can begin discussing the technical implementation of module management/maintenance for zim core.

**Comment by Plloi** [comment]
I see the module repos are at present just their current heads...
It might be worthwhile to blow them out and rebuild them from the original zim repo with filter-branch so they retain their change history. 

This isn't necessary of course, but if it's going to be done... sooner is way better than later

**Comment by Eriner** [comment]
The majority of modules have undergone very little change even from zim's inception (for many, it has been a matter of updating an external submodule), so I'm not sure if it's really worth extrapolating those commits; the credit to the authors will remain in the core repository when it's transferred, so I'm not too worried about that. 

I think it may be a good idea to keep them as "clean" as possible after the transfer. No longer do we have to use the `[git] fixed a thing` commit style. Also, many of the commit messages would lose their context, specifically referring to/closing issues that existed in the main repository that no longer exist in the submodule repos. 

If anyone has any arguments for transferring the old commits, I'm more than happy to discuss it.

**Comment by ericbn** [comment]
I agree that we could have all the history in the main zimframework/zim repo, with a commit that deletes the modules on top of everything. And maybe the original Eriner/zim issues could be kept for historical purposes.

**Comment by dvcrn** [comment]
@Eriner I think you could just rename this repository here to `zim-legacy-issues` (or something like that) and then move the Eriner/zim into to the organization. That should retain the issues from both repos 😄

**Comment by N4M3Z** [comment]
Hello everyone, I have recently revisited my allegiance to Prezto and I discovered the renewed endeavor behind this project. I really like the idea of separating the modules into different repositories as I believe Prezto failed for it's inability to allow people to contribute modules that would make their way into the project (I have submitted multiple PRs to Prezto myself, I believe none of them have been merged). Since this new way of organization naturally paves a way towards module managers like zgen and nowadays zplug, I wonder whether you considered simply joining forces with zplug instead of building a loading system of your own?

**Comment by Eriner** [comment]
@dvcrn if you look at the issues in the top of this thread, you can see how I ported the old issues. I'll likely do the same for the issues here.

@N4M3Z zgen and zplug are extremely bloated when compared to zim-core (the module loading code [here](https://github.com/Eriner/zim/blob/master/init.zsh)), which is all of ~40 lines. I understand that code length isn't everything, but I've also looked at the speed of zplug, and tested the speed with [zsh-framework-benchmark](https://github.com/Eriner/zsh-framework-benchmark). Unfortunately, the zplug test is currently not functioning and it's been about a year since I've updated the [speed comparison page, which includes zplug](https://github.com/Eriner/zim/wiki/Speed), but you get the idea.

All we need is something minimal, fast, and clean. We don't need an all-in-one-magical-unicorn loader, if you catch my drift.

**Comment by Eriner** [comment]
@ericbn That was the plan; transfer Eriner/zim to zimframework/zim, branch to testing, rm modules from zimframework/zim, develop loading system/module management code to load zimframework/${module_name} modules AND external modules, test and confirm clean upgrade path, merge to master and release v1.0.

**Comment by Eriner** [comment]
Module migration to organization is complete, PR to migrate the core zim repo to use these modules is here: https://github.com/zimframework/zim/pull/6

**Comment by paulmelnikow** [comment]
Hi all, I’ve recently signed on as a maintainer to the Prezto community fork. In the process of getting myself up to speed, I’ve been researching alternatives, and wow! I'm really excited about Zim. @Eriner, you’ve done such awesome work!

Glad to find this repo and learn you’re thinking about maintainability. Would be great to find some good ways to work together, and to combine forces where we can.

Wanted to join this discussion and suggest a slight twist on the plan. Sorry for being a bit forward here. I’m newly introduced to the Zim community and it may not be my place. It's offered with the utmost of respect for the work and investment you have all made, including the migration in progress.

> I believe Oh My Zsh and even Prezto failed to grow in an organized fashion mainly because they lack incorporating some kind of plugin system that would allow people to contribute with external modules and keep the core clean and simple.

^^ @ericbn hits the nail on the head.

Building on this idea, what do you all think about conceiving Zim as _a tight set of core modules_?

These core modules have broad impact. There’s significant benefit from maintaining them together, both for the developer, and for providing a cohesive user experience. The modules may have hard coupling as well. They may define interfaces for, or manage integration with, external plugins.

As you’ve observed, external plugin loading is a significant challenge of its own. I found in my research a good handful of _concurrently active plugin-loader projects_. In addition to zplug and antigen, there’s a project called [zulu](https://github.com/zulu-zsh/zulu), which is setting out to be a plugin repository, plugin manager, and framework. There is [zpm](https://github.com/zpm-zsh/zpm). There is [antibody](https://github.com/getantibody/antibody), a second project called [zpm](https://github.com/zpm-project/zpm-zsh), [antigen-hs](https://github.com/Tarrasch/antigen-hs), and [zgen](https://github.com/tarjoilija/zgen) — each of which is aiming to be a faster zplug. Almost all of the projects I've named have been active in the last few weeks.

My suggestion here is to focus on what is strongest about Zim, which is that it’s the _fastest, most maintainable, and best loved core_. Was chatting with @facastagnini today and we’re both excited about this. What was once great about Prezto is what is now great about Zim.

By not trying to be all things to all people, you can defer the complexity of building a generic module loader. Instead you can let the proliferation of plugin managers play out. I predict that eight (or even four) months from now things may start to converge a a bit. Some of these tools will mature, traction will accelerate, plugin interoperability will improve, independent plugins will themselves mature. Maybe a registry will take off, which would change the landscape a lot.

The external plugins could be managed _through a core module_. Perhaps the Zim core developers could select one of them to start, and swap it out down the line when there’s a better alternative. You could be agnostic, providing recipes. You could bundle in a couple and let the user choose.

It seems like the core doesn't need an external module loader _immediately_. If you're willing to trust that a good one will come within a few months, and can prune out non-essentials like `archive` and `fasd`, perhaps it would let you run Zim core as a monorepo, which would make issue tracking and pull requests easier to manage, reducing overhead for developers.

Realize these thoughts may be a few weeks late, though I’ve been steeping in this for a few days and wanted to offer them up. Hope they're taken in the constructive spirit in which they're offered!

**Comment by ericbn** [comment]
@paulmelnikow, I'm happy to ready you saying that my comment "hits the nail on the head". 😄 I also like Zim being the "_fastest, most maintainable, and best loved core_". And I would also add that we're trying to keep Zim **simple**. (Prezto, for example, has too many fine grained settings, getting to the point of having setting names as long as ":prezto:module:editor:info:keymap:primary:insert", really!)

@Eriner, correct me if I'm wrong here:

The plan is to separate all the modules into independent repos already. You called them "core modules", but just to keep naming consistent we're calling "core" the code that manages the modules. This "core" code is in `.zlogin`, `.zim/init.zsh`, `.zim/tools/*`, and also in the `debug` and `meta` modules. The "core" code is responsible for:
- pre-zcompiling the modules code (with further investigation to be done in https://github.com/Eriner/zim/issues/86 and https://github.com/zimframework/zim/issues/5)
- loading the modules listed in `zmodules`
- providing the ability to
    - manage the zcompiled files (rebuild and clean cache)
    - update the zim and modules source code
    - generate info and debug data for users wishing to open an issue

I must confess I never used a Zsh plugin manager (as I've only been through Oh-My-Zsh, Prezto and now Zim). Can we outsource the "loading the modules" and "update the modules source code" to an external plugin manager that would do just that? If we exclude Zulu (that seems to be doing lots of things, and even depends on having the modules registered on its own index repo), and plugin managers that offer extra functionality or the ability to automatically have Oh-My-Zsh or Prezto as sources, do we have any left?

**Comment by Eriner** [comment]
> Can we outsource the "loading the modules" and "update the modules source code" to an external plugin manager that would do just that?

Maybe. I appreciate and agree with @paulmelnikow's description of the situation. I'm not sure what the best course of action is. If we offload the loading of modules (universally, abandoning the "core" loading code) and adopt an upstream provider(s), that means we have no control of how/what loads, and we're at their mercy for load time. If there's something out there that does what we want efficiently, quickly, and cleanly, I'm more than willing to adopt it rather than reinvent the wheel, but it must fall in line with these principles - I'm open to suggestions and possibilities. Zplug, when I last tested (which was indeed a while ago, so things may have changed) with the same modules and zim, takes *significantly* longer to load: https://github.com/Eriner/zim/wiki/Speed. The speeds were even worse than prezto's speed, which was one of the deficiencies that the creation of zim spawned from.

Alternatively, if we persue the mantra of "fastest, most maintainable, and best loved core", in my eyes this means **zim** loads the modules, but provides a structure (individual repos, for example) for other managers to pull plugins where desired.

What do you guys think?

**Comment by sodiumjoe** [comment]
Just to throw in my two cents, I also recently (within the last week) tried to switch from prezto to zplug (and a bunch of other plugin managers) and ended up returning to prezto because it was the fastest. Until I found Zim! It doesn't seem like a good time to try to adopt another plugin manager and I think the vision that you've laid out is extremely compelling.

I'm really excited to see how the user configuration and external module stuff plays out, which is the only thing keeping me from adopting Zim completely.

Would it be possible to modularize the plugin loading system such that it could be swapped out at a later time, once the zsh plugin manager war dust has settled?

**Comment by ericbn** [comment]
I've just run the benchmark in my machine (had to run in a Ubuntu VirtualBox and remove the "spin" code, as it was hanging):

#### MacBook Pro (SSD, Intel Core i7 @2.2GHz)
```
The average startup time for vanilla is: 0.038124999999999937
The average startup time for oh-my-zsh is: 0.13566250000000019
The average startup time for prezto is: 0.45254249999999935
The average startup time for zim is: 0.090450000000000141
The average startup time for zplug is: 0.005549999999999995
```

Looks like zplug improved a lot since [last benchmarks](https://github.com/Eriner/zim/wiki/Speed). @paulmelnikow, any other plugin manager worth testing?

**Comment by paulmelnikow** [comment]
Will respond to earlier comments when I have a moment, but wanted to say that zplug number is suspiciously low. It's lower than vanilla, which is impossible.

Just found [the benchmark code](https://github.com/Eriner/zsh-framework-benchmark). That script is super useful! Would love to get this into CI. Maybe I can help with that.

[The line that installs zplug](https://github.com/Eriner/zsh-framework-benchmark/blob/master/frameworks/zplug.zsh#L6) is hitting https://git.io/zplug which redirects to a 404.

For some reason I can't get the benchmarks to run on my system. The script gets part way through vanilla and then appears to stop. My prompt comes back. Same problem in my own shell, a clean Prezto install, and an account with a bash shell.

**Comment by paulmelnikow** [comment]
> I would also add that we're trying to keep Zim **simple**.

Yay for simple!

My bad about misusing the terminology. Makes sense that “core” means the module loader, give or take. It’s that plus the “basic modules” @ericbn enumerated in the top post, and what @Eriner called the “essential modules” in #1, that I was referring to when I said _"best loved ~~core~~."_

> If we offload the loading of modules (universally, abandoning the "core" loading code) and adopt an upstream provider(s), that means we have no control of how/what loads, and we're at their mercy for load time. If there's something out there that does what we want efficiently, quickly, and cleanly, I'm more than willing to adopt it rather than reinvent the wheel, but it must fall in line with these principles - I'm open to suggestions and possibilities.

That makes sense. To be honest, though, I don’t think any of the plugin managers are up to snuff. None are anywhere close to tackling loading as well as you are right now. Some of them index the modules – i.e., decide what to load – and cache the result. They do this quite well, I might add. None of them are zcompiling the modules, though one of them (zulu) is zcompiling itself. Your loader is a big part of the “best loved essentials” I’m praising. Please don’t abandon it. 😃 

> Would it be possible to modularize the plugin loading system such that it could be swapped out at a later time, once the zsh plugin manager war dust has settled?

That ^^ was @sodiumjoe, and is my suggestion in a nutshell.

Put another way, let’s decouple two pieces of the functionality for external code:

1. The installation and updating of external module source code, configuration of which modules are active, and generation of an index of active modules (in a form TBD). _This is the responsibility of the plugin manager._
2. The loading of those modules. _This is the responsibility of zim “core.”_

Here's why. The Zim “core” code as it is today, and this team, are absolutely best equipped to do the second. By far. At least in my estimation. Sure, you’ll need to build in some functionality to handle an index of active modules, which may be technically challenging. However, I don’t think any of the plugin-manager projects have reached nearly the level of sophistication you have when it comes to loading. And reasonably so. They haven’t focused on performance as much as you have.

It could also be important for Zim to provide “essential modules” that provide good defaults out of the box. If you were to decouple loading from plugin installation, you’d have some options for handling these “essential” or “internal” modules. You could entrust them to the plugin manager. You could handle them as you are now – using git. You could wait and see.

What do you think?

By the way, I finally got a chance to install and play with Zim yesterday. My initial impressions were good. Blazing startup. Just wow! Stuff generally worked, including my homegrown module loader. I ran into a couple hiccups interoperating with other frameworks – that is, it took me a while to be able to switch cleanly between Prezto and Zim. Interop may be a factor for other people who are considering / in process of migrating. I do prefer to keep my config outside the framework repo. (A discussion for another thread!) Now that I’m depending on a plugin manager (and I expect there will be one in Prezto soon) that’s not only a key feature but an interop concern as well.

**Comment by Eriner** [comment]
@ericbn thanks for the benchmarks, glad to see zim is still the speediest! :)

> Just found the benchmark code. That script is super useful! Would love to get this into CI. Maybe I can help with that.

@paulmelnikow I'd be interested in integrating automated drone tests, sounds like a useful functionality (to update benchmarks weekly, for example. I can hook into one of my drone server clusters to handle this.)

> The installation and updating of external module source code, configuration of which modules are active, and generation of an index of active modules (in a form TBD). This is the responsibility of the plugin manager.

This is the elephant in the room, and I would be more than happy to delegate this to some other code that I don't have to write, maintain, or even care about. The *zim* loading itself, as you described, I would like to keep within zim so that you can simply `source` or `load 'zimframework/zim'` or something similar without loosing the optimizations and speed. The new structure should allow the main modules to be loaded (and upgraded and managed) by other frameworks, which is good, but the internal modules should be initialized and loaded by a single module/loader (core). This allows us explicit control over how/what is loaded, and lets us maintain our own optimizations.

Thoughts?
